File: jenkins/jenkinsEC2/cdk/lib/ecs.ts
import { aws_servicediscovery as sd, aws_ecs as ecs, aws_ec2 as ec2, aws_autoscaling as autoscaling, aws_efs as efs } from 'aws-cdk-lib';
import { Stack, StackProps } from 'aws-cdk-lib';
import { Construct } from 'constructs';

interface EcsProps extends StackProps {
  vpc: ec2.IVpc,
  serviceDiscoveryNamespace: string,
}

export class Ecs extends Stack {
  public readonly cluster: ecs.Cluster;
  public readonly asg: autoscaling.AutoScalingGroup;
  public readonly efsSecGrp: ec2.SecurityGroup;
  public readonly efsFilesystem: efs.CfnFileSystem;

  constructor(scope: Construct, id: string, props: EcsProps) {
    super(scope, id, props);

    const serviceDiscoveryNamespace = props.serviceDiscoveryNamespace;
    const vpc = props.vpc;

    // ECS Cluster
    this.cluster = new ecs.Cluster(this, "EcsCluster", {
      clusterName: 'jenkins-production',
      vpc: vpc,
      defaultCloudMapNamespace: {
        name: serviceDiscoveryNamespace,
        type: sd.NamespaceType.DNS_PRIVATE,
      }
    });

    // EC2
    const asg = this.cluster.addCapacity("Ec2", {
      instanceType: new ec2.InstanceType('t3.xlarge'),
      keyName: "jenkins-prod-key",
      allowAllOutbound: true,
      associatePublicIpAddress: true,
      vpcSubnets: { subnetType: ec2.SubnetType.PUBLIC }
    });

    // EFS Security Group
    const efsSecGrp = new ec2.SecurityGroup(this, "EFSSecGrp", {
      securityGroupName: "jenkins-prod-efs-sg",
      vpc: vpc,
      allowAllOutbound: true,
    });
    efsSecGrp.addIngressRule(
        this.cluster.connections.securityGroups[0],
        new ec2.Port({
          protocol: ec2.Protocol.ALL,
          stringRepresentation: "ALL",
          fromPort: 2049,
          toPort: 2049,
        }),
        "EFS"
    );

    // EFS Filesystem
    const efsFilesystem = new efs.CfnFileSystem(this, "EFSBackend");
    vpc.privateSubnets.forEach((subnet, idx) => {
      new efs.CfnMountTarget(this, `EFS${idx}`, {
        fileSystemId: efsFilesystem.ref,
        subnetId: subnet.subnetId,
        securityGroups: [efsSecGrp.securityGroupId]
      });
    });

    // User Data
    const userData = `
sudo yum install -y amazon-efs-utils
sudo mkdir /mnt/efs
sudo chown -R ec2-user: /mnt/efs
sudo chmod -R 0777 /mnt/efs
sudo mount -t efs -o tls /mnt/efs ${efsFilesystem.ref}:/ efs`;
    asg.addUserData(userData);
  }
}

File: jenkins/JenkinsEC2Py/cdk/jenkins/ecs.py
from aws_cdk import (
    aws_ecs as ecs,
    aws_ec2 as ec2,
    aws_efs as efs,
    Stack
)
from constructs import Construct
from configparser import ConfigParser

config = ConfigParser()
config.read('config.ini')


class ECSCluster(Stack):
    def __init__(self, scope: Construct, id: str, vpc, service_discovery_namespace, **kwargs):
        super().__init__(scope, id, **kwargs)

        self.vpc = vpc
        self.service_discovery_namespace = service_discovery_namespace

        # Create ECS cluster
        self.cluster = ecs.Cluster(
            self, "ECSCluster",
            vpc=self.vpc,
            default_cloud_map_namespace=ecs.CloudMapNamespaceOptions(name=service_discovery_namespace)
        )

        if config['DEFAULT']['ec2_enabled'] == "yes":
            self.asg = self.cluster.add_capacity(
                "Ec2",
                instance_type=ec2.InstanceType(config['DEFAULT']['instance_type']),
                key_name="jenkinsonaws",
                # ... other properties ...
            )

            self.efs_sec_grp = ec2.SecurityGroup(
                self, "EFSSecGrp",
                vpc=self.vpc,
                allow_all_outbound=True,
            )

            self.efs_sec_grp.add_ingress_rule(
                peer=ec2.Peer.security_group_id(self.asg.connections.security_groups[0].security_group_id),
                connection=ec2.Port.tcp(2049),
                description="EFS"
            )

            self.efs_filesystem = efs.CfnFileSystem(
                self, "EFSBackend",
            )

            counter = 0
            for subnet in self.vpc.private_subnets:
                efs.CfnMountTarget(
                    self, f"EFS{counter}",
                    file_system_id=self.efs_filesystem.ref,
                    subnet_id=subnet.subnet_id,
                    security_groups=[self.efs_sec_grp.security_group_id]
                )
                counter += 1

            self.user_data = """
                sudo yum install -y amazon-efs-utils
                sudo mkdir /mnt/efs
                sudo chown -R ec2-user: /mnt/efs
                sudo chmod -R 0777 /mnt/efs
                sudo mount -t efs -o tls /mnt/efs {}:/ efs
                """.format(self.efs_filesystem.ref)

            self.asg.add_user_data(self.user_data)
